apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.GenerateOverloads/Program.cs

```diff
@@ -0,0 +1,119 @@
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+
+namespace Google.Cloud.BigQuery.V2.GenerateOverloads
+{
+    /// <summary>
+    /// BigQuery-specific overload generator.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The BigQuery library ends up with a lot of very similar methods, varying by how a dataset/job/table is specified,
+    /// and whether or not it's asynchronous. This generator allows the description to be written a single time and then used
+    /// to generate all the overloads. XML is chosen as the file format as most of the definition is in terms of XML documentation.
+    /// </para>
+    /// <para>
+    /// For ease of later regeneration, files are checked into source control. The generation process is still run manually though.
+    /// It can be rerun for the same source file, and the code will be replaced using region markers to indicate where methods are defined.
+    /// If manual changes are required, they will need to be reperformed each time the generator is run. Ideally, make the generator smarter instead.
+    /// TODO: Have some way of making the tool prompt for this...
+    /// </para>
+    /// <para>
+    /// This generator only handles the async + entity referencing overloads. Where multiple overloads have different parameter types
+    /// (e.g. UploadJson with Stream and String[] parameters) these would be generated from multiple XML files.
+    /// </para>
+    /// </remarks>
+    class Program
+    {
+        private static int Main(string[] args)
+        {
+            if (args.Length != 2)
+            {
+                Console.WriteLine("Arguments: <xml filename> <source file>");
+                return 1;
+            }
+
+            try
+            {
+                XDocument methodDoc = XDocument.Load(args[0]);
+                ApiMethod method = new ApiMethod(methodDoc);
+                var existingCode = File.ReadLines(args[1]);
+                var newCode = ModifyCode(method, existingCode);
+                File.WriteAllLines(args[1], newCode);
+                return 0;
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Error: {e}");
+                return 1;
+            }
+        }
+
+        private static IEnumerable<string> ModifyCode(ApiMethod method, IEnumerable<string> existingCode)
```

Comment:
I'd be happier with code-generation that completely generates and replaces the source file.
(In my opinion) this would be cleaner, and be less likely to produce problems in the future.
But I can see that'll be more complex, so this is fine. Could be changed later if desirable.